dask_jobqueue/core.py

```diff
@@ -1,23 +1,73 @@
+from __future__ import absolute_import, division, print_function
+
 import logging
+import math
 import os
 import shlex
 import socket
 import subprocess
 import sys
+from collections import OrderedDict
 from contextlib import contextmanager
 
 import dask
 import docrep
 from distributed import LocalCluster
 from distributed.deploy import Cluster
-from distributed.utils import get_ip_interface, ignoring, parse_bytes, tmpfile
+from distributed.utils import get_ip_interface, parse_bytes, tmpfile
+from distributed.diagnostics.plugin import SchedulerPlugin
 
 dirname = os.path.dirname(sys.executable)
 
 logger = logging.getLogger(__name__)
 docstrings = docrep.DocstringProcessor()
 
 
+def _job_id_from_worker_name(name):
+    ''' utility to parse the job ID from the worker name'''
+    return name.split('-')[-1]
+
+
+class Job(object):
+    def __init__(self, job_id, workers=None, status=None):
+        self.job_id = job_id
+        self.status = status
+        if workers is None:
+            workers = []
+        self.workers = workers
+
+    def update(self, worker=None, status=None):
+        if worker is not None:
+            self.workers.append(worker)
+        if status is not None:
+            self.status = status
+
+    def __repr__(self):
+        return "<%s: %r, status: %r, workers: %r>" % (
+            self.__class__.__name__, self.job_id, self.status, self.workers)
+
+
+class JobQueuePlugin(SchedulerPlugin):
+    def __init__(self):
+        self.pending_jobs = OrderedDict()
+        self.running_jobs = OrderedDict()
+        self.finished_jobs = OrderedDict()
+
+    def add_worker(self, scheduler, worker=None, name=None, **kwargs):
+        job_id = _job_id_from_worker_name(worker.name)
+        if job_id not in self.running_jobs:
+            self.running_jobs[job_id] = self.pending_jobs.pop(job_id)
+        self.running_jobs[job_id].update(worker=worker, status='running')
+
+    def remove_worker(self, scheduler=None, worker=None, **kwargs):
+        job_id = _job_id_from_worker_name(worker.name)
+        if job_id not in self.finished_jobs:
+            self.finished_jobs[job_id] = self.running_jobs.pop(job_id)
+        self.finished_jobs[job_id].update(status='finished')
+        if self.finished_jobs[job_id].workers:
+            self.finished_jobs[job_id].workers = []
```

Comment:
If I understand correctly, you are considering the workers are ending all at once, when the first worker corresponding to this job_id is removed. Perhaps we could remove workers one by one, just to be sure? This may be overkill.

Reply:
Good idea. This was a pretty easy fix so I have included it in 4776892.